lib/go-tc/v13/origins.go

```diff
@@ -0,0 +1,150 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+type OriginsResponse struct {
+	Response []OriginNullable `json:"response"`
+}
+
+// OriginDetailResponse is the JSON object returned for a single origin
+type OriginDetailResponse struct {
+	Response OriginNullable `json:"response"`
+	tc.Alerts
+}
+
+type Origin struct {
+	Cachegroup        string       `json:"cachegroup" db:"cachegroup"`
+	CachegroupID      int          `json:"cachegroupId" db:"cachegroup_id"`
+	Coordinate        string       `json:"coordinate" db:"coordinate"`
+	CoordinateID      int          `json:"coordinateId" db:"coordinate_id"`
+	DeliveryService   string       `json:"deliveryService" db:"deliveryservice"`
+	DeliveryServiceID int          `json:"deliveryServiceId" db:"deliveryservice_id"`
+	FQDN              string       `json:"fqdn" db:"fqdn"`
+	ID                int          `json:"id" db:"id"`
+	IP6Address        string       `json:"ip6Address" db:"ip6_address"`
+	IPAddress         string       `json:"ipAddress" db:"ip_address"`
+	IsPrimary         bool         `json:"isPrimary" db:"is_primary"`
+	LastUpdated       tc.TimeNoMod `json:"lastUpdated" db:"last_updated"`
+	Name              string       `json:"name" db:"name"`
+	Port              int          `json:"port" db:"port"`
+	Profile           string       `json:"profile" db:"profile"`
+	ProfileID         int          `json:"profileId" db:"profile_id"`
+	Protocol          string       `json:"protocol" db:"protocol"`
+	Tenant            string       `json:"tenant" db:"tenant"`
+	TenantID          int          `json:"tenantId" db:"tenant_id"`
+}
+
+type OriginNullable struct {
+	Cachegroup        *string       `json:"cachegroup" db:"cachegroup"`
+	CachegroupID      *int          `json:"cachegroupId" db:"cachegroup_id"`
+	Coordinate        *string       `json:"coordinate" db:"coordinate"`
+	CoordinateID      *int          `json:"coordinateId" db:"coordinate_id"`
+	DeliveryService   *string       `json:"deliveryService" db:"deliveryservice"`
+	DeliveryServiceID *int          `json:"deliveryServiceId" db:"deliveryservice_id"`
+	FQDN              *string       `json:"fqdn" db:"fqdn"`
+	ID                *int          `json:"id" db:"id"`
+	IP6Address        *string       `json:"ip6Address" db:"ip6_address"`
+	IPAddress         *string       `json:"ipAddress" db:"ip_address"`
+	IsPrimary         *bool         `json:"isPrimary" db:"is_primary"`
+	LastUpdated       *tc.TimeNoMod `json:"lastUpdated" db:"last_updated"`
+	Name              *string       `json:"name" db:"name"`
+	Port              *int          `json:"port" db:"port"`
+	Profile           *string       `json:"profile" db:"profile"`
+	ProfileID         *int          `json:"profileId" db:"profile_id"`
+	Protocol          *string       `json:"protocol" db:"protocol"`
+	Tenant            *string       `json:"tenant" db:"tenant"`
+	TenantID          *int          `json:"tenantId" db:"tenant_id"`
+}
+
+type OriginBuilder struct {
```

Comment:
Yeah they're not currently used anywhere in the codebase, but I found it useful when I was implementing a test program using the Origin client. Since none of the other client implementations use the `*Nullable` structs yet, something like this hasn't really been necessary. Without it, you basically have to declare a temporary variable and `&` that variable when setting the pointer in the `*Nullable` struct. Using this Builder pattern you basically just:
```
foo := NewOriginBuilder().CacheGroupID(1).CoordinateID(1).Build()
```
as opposed to:
```
cgID := 1
coordID := 1
foo := OriginNullable{CachegroupID: &cgID, CoordinateID: &coordID}
```

Reply:
I'm a -1 on this. I can see the convenience, but it's very non-standard. The benefit isn't worth the readability cost, for new devs to figure out.

To avoid multiple lines, the Go idiom is to have a `NewOrigin()` func, that either takes values to set, or takes no parameters and initializes all the pointers.

I'm guessing most usages will be loading from SQL and parsing JSON from POSTs, so it probably won't be used enough to justify the learning curve, either.

Reply:
Yeah I don't feel super strongly about this, just found it useful. Dylan mentioned that he sometimes uses a `func IntPointer(i int) *int {return &i}` to alleviate some of the headache with this, which might be better than maintaining this builder pattern here. I can take it out 👍